tools/mpremote: Add `delay` and `rtc` commands. Extend documentation. (#11659)
micropython/micropython: docs/reference/mpremote.rst

====================================

The ``mpremote`` command line tool provides an integrated set of utilities to
remotely interact with, manage the filesystem on, and automate a MicroPython
device over a serial connection.

To use mpremote install it via ``pip``:

.. code-block:: bash

    $ pip install --user mpremote

Another option is to install or run it using `pipx <https://pypa.github.io/pipx/>`_:

.. code-block:: bash

>     $ pipx run mpremote

Author: I've updated it, see what you think.

  device. Strictly, the ``repl`` command is just functioning as a terminal
  (or "serial monitor") to access the device. Because this command does not
  trigger the :ref:`auto-reset behavior <mpremote_reset>`, this means that if
  a program is currently running, you will first need to interrupt it with
  ``Ctrl-C`` to get to the REPL, which will then allow you to access program
  state. You can also use ``mpremote soft-reset repl`` to get a "clean" REPL
  with all program state cleared.

.. _mpremote_command_eval:

- **eval** -- evaluate and print the result of a Python expression:

  .. code-block:: bash

      $ mpremote eval <string>

.. _mpremote_command_exec:

- **exec** -- execute the given Python code:

  .. code-block:: bash

      $ mpremote exec <string>

  By default, ``mpremote exec`` will display any output from the expression until it
  terminates. The ``--no-follow`` flag can be specified to return immediately and leave
  the device running the expression in the background.

.. _mpremote_command_run:

- **run** -- run a script from the local filesystem:

  .. code-block:: bash

      $ mpremote run <file.py>

>   This will copy the file to RAM on the device and execute it directly.

Author: I'm not sure about the suggested wording because to me "transfer to the device" implies copying it to the filesystem. I've updated the wording to not say "copy". (And sentence about what this command is useful for).

  By default, ``mpremote run`` will display any output from the script until it
  terminates. The ``--no-follow`` flag can be specified to return immediately and leave
  the device running the script in the background.

.. _mpremote_command_fs:

- **fs** -- execute filesystem commands on the device:

  .. code-block:: bash

>       $ mpremote fs <sub-command>

Author: I am confused by this too :p

I've added a `Note:` to this section, to at least acknowledge that this is a thing and not some extra magic.

  variable ``$EDITOR``). If the editor exits successfully, the updated file will
  be copied back to the device.

.. _mpremote_command_mip:

- **mip** -- install packages from :term:`micropython-lib` (or GitHub) using the ``mip`` tool:

  .. code-block:: bash

      $ mpremote mip install <packages...>

  See :ref:`packages` for more information.

.. _mpremote_command_mount:

- **mount** -- mount the local directory on the remote device:

  .. code-block:: bash

>       $ mpremote mount [options] <local-dir>

Reviewer: This is a good improvement - for a very difficult topic to explain! 

I wonder if this should have a short animated gif? To show a simple sequence like:

- ls device - empty
- ls PC - foo.py
- mount, ls device - foo.py

Or maybe we just need to record a video, as we've discussed.

Author: Yep. +1 to video.

  This allows the remote device to see the local host directory as if it were
  its own filesystem. This is useful for development, and avoids the need to
  copy files to the device while you are working on them.

  The device installs a filesystem driver, which is then mounted in the device

>   VFS as ``/remote``, which uses the serial connection to ``mpremote`` as a

Reviewer: Reference to VFS may be needed?

  side-channel to access files. The device will have its current working
  directory (via ``os.chdir``) set to ``/remote`` so that imports and file

>   access will go there in place on the default filesystem while the mount is

Reviewer: ```suggestion
  access will occur there instead of the default filesystem while the mount is
```

soft-reset for the first time, it will not be done again automatically, until a
``disconnect`` command is issued.

Auto-soft-reset behaviour can be controlled by the ``resume`` command. This
might be useful to use the ``eval`` command to inspect the state of of the
device.  The ``soft-reset`` command can be used to perform an explicit soft
reset in the middle of a sequence of commands.

.. _mpremote_shortcuts:

Shortcuts
---------

Shortcuts can be defined using the macro system.  Built-in shortcuts are:

> - ``devs``: Alias for ``connect list``

Reviewer: FWIW, I'm often asked how `devs` filters the serial device list to detect MicroPython devices.

Author: `devs` (i.e.`connect list`) doesn't filter, but `auto` does, so I have explained that.

  mpremote

> This automatically connects to the first available device and implicitly runs
> the ``repl`` command.

Reviewer: Consider making these more succinct by dropping "This..."

```suggestion
Connect to the first available device and implicitly run the ``repl`` command.
```

.. code-block:: bash

  mpremote a1

This connects to the device at ``/dev/ttyACM1`` and implicitly runs the ``repl``
command.

.. code-block:: bash

  mpremote connect /dev/ttyUSB0

This allows you to explicitly specify which device to connect to, and as above,
implicitly runs the ``repl`` command.

.. code-block:: bash

  mpremote a1 ls

> This connects to the device at ``/dev/ttyACM0`` and then runs the ``ls`` command.

Reviewer: Consider using `c0` in one of these examples to reiterate that `mpremote` works on Windows?
